File: LeetCodePractice/src/greedy/MajorityElementII.java
package greedy;
/*
Given an integer array of size n, find all elements that appear more than ⌊ n/3 ⌋ times.

Note: The algorithm should run in linear time and in O(1) space.

Example 1:

Input: [3,2,3]
Output: [3]
Example 2:

Input: [1,1,1,3,3,2,2,2]
Output: [1,2]
*/

import java.util.ArrayList;
import java.util.List;

public class MajorityElementII {
    /**
     * @param nums: A list of integers
     * @return: The majority numbers that occurs more than 1/3
     */
    public List<Integer> majorityElement(int[] nums) {
        // write your code
        List<Integer> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return res;
        }
        int count1 = 0;
        int count2 = 0;
        int candidate1 = -1;
        int candidate2 = -1;
        int len = nums.length;
        for (int i = 0; i < len; i++) {
            if (count1 == 0) {
                candidate1 = nums[i];
                count1 = 1;
            } else if (candidate1 == nums[i]) {
                count1++;
            } else if (count2 == 0) {
                candidate2 = nums[i];
                count2 = 1;
            } else if (candidate2 == nums[i]) {
                count2++;
            } else {
                count1--;
                count2--;
            }
        }
        count1 = 0;
        count2 = 0;
        for (int i = 0; i < len; i++) {
            if (candidate1 == nums[i]) {
                count1++;
            } else if (candidate2 == nums[i]) {
                count2++;
            }
        }
        if (count1 > len / 3) {
            res.add(candidate1);
        }
        if (count2 > len / 3) {
            res.add(candidate2);
        }
        return res;
    }
}

File: LeetCodePython/MajorityElementII.py
"""
Given an integer array of size n, find all elements that appear more than ⌊ n/3 ⌋ times.

Example 1:
Input: nums = [3,2,3]
Output: [3]

Example 2:
Input: nums = [1]
Output: [1]

Example 3:
Input: nums = [1,2]
Output: [1,2]


Constraints:
1 <= nums.length <= 5 * 10^4
-109 <= nums[i] <= 10^9

Follow up: Could you solve the problem in linear time and in O(1) space?
"""
from typing import List


class MajorityElementII:
    def majorityElement(self, nums: List[int]) -> List[int]:
        if not nums:
            return []
        cnt1, cnt2, cand1, cand2 = 0, 0, 0, 1
        for n in nums:
            if n == cand1:
                cnt1 += 1
            elif n == cand2:
                cnt2 += 1
            elif cnt1 == 0:
                cand1, cnt1 = n, 1
            elif cnt2 == 0:
                cand2, cnt2 = n, 1
            else:
                cnt1 -= 1
                cnt2 -= 1
        return[n for n in (cand1, cand2) if nums.count(n) > len(nums) // 3]